File: solutions/python/25.py
def execute(code: str) -> int:
    value = 0
    instruction_index = 0
    loop_index = 0
    while (instruction_index < len(code)):
        instruction = code[instruction_index]
        if (instruction == "+"):
            value += 1
        elif (instruction == "-"):
            value -= 1
        elif (instruction == "["):
            loop_index = instruction_index
            if (value == 0):
                instruction_index = code.find("]", instruction_index)
        elif (instruction == "]" and value != 0):
            instruction_index = loop_index
        elif (instruction == "{" and value == 0):
            instruction_index = code.find("}", instruction_index)
        instruction_index += 1
    return value


if __name__ == "__main__":
    code_1 = execute("+++")  # 3
    print(code_1)

    code_2 = execute("+--")  # -1
    print(code_2)

    code_3 = execute(">+++[-]")  # 0
    print(code_3)

    code_4 = execute(">>>+{++}")  # 3
    print(code_4)

    code_5 = execute("+{[-]+}+")  # 2
    print(code_5)

    code_6 = execute("{+}{+}{+}")  # 0
    print(code_6)

    code_7 = execute("------[+]++")  # 2
    print(code_7)

    code_8 = execute("-[++{-}]+{++++}")  # 5
    print(code_8)

"""
  Score: ★★★★★
"""

File: solutions/js/25.js
/**
 * @param {string} code - The magical program to execute
 * @returns {number} - The final value after executing the program
 */
function execute(code) {
  let value = 0;
  let instructionIndex = 0;
  let loopIndex = 0;
  while (instructionIndex < code.length) {
    const instruction = code[instructionIndex];
    if (instruction === "+") {
      value++;
    } else if (instruction === "-") {
      value--;
    } else if (instruction === "[") {
      loopIndex = instructionIndex;
      if (value === 0) {
        instructionIndex = code.indexOf("]", instructionIndex);
      }
    } else if (instruction === "]" && value !== 0) {
      instructionIndex = loopIndex;
    } else if (instruction === "{" && value === 0) {
      instructionIndex = code.indexOf("}", instructionIndex);
    }
    instructionIndex++;
  }
  return value;
}

const code1 = execute("+++"); // 3
console.log(code1);

const code2 = execute("+--"); // -1
console.log(code2);

const code3 = execute(">+++[-]"); // 0
console.log(code3);

const code4 = execute(">>>+{++}"); // 3
console.log(code4);

const code5 = execute("+{[-]+}+"); // 2
console.log(code5);

const code6 = execute("{+}{+}{+}"); // 0
console.log(code6);

const code7 = execute("------[+]++"); // 2
console.log(code7);

const code8 = execute("-[++{-}]+{++++}"); // 5
console.log(code8);

/* 
  Score: ★★★★★
*/
